File: Exercicios_diarios/Revisão_intermediaria/Ciclo de Lista Vinculada II.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def detectCycle(self, head):
        lento = rapido = head
        while rapido and rapido.next:
            lento, rapido = lento.next, rapido.next.next
            if lento == rapido:
                break
        else:
            return None
        while head != lento:
            head, lento = head.next, lento.next
        return head


class Solution:
    def detectCycle(self, head):
        lento, rapido = head, head
        while rapido and rapido.next:
            lento = lento.next
            rapido = rapido.next.next
            if lento == rapido:
                while head != rapido:
                    head = head.next
                    rapido = rapido.next
                return head
        return None

File: Python/Exercicios_diarios/Revisão_intermediaria/Ciclo de Lista Vinculada II.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        ListNode *lento = head, *rapido = head;
        while (rapido && rapido->next)
        {
            lento = lento->next;
            rapido = rapido->next->next;
            if (lento == rapido)
                break;
        }
        if (!(rapido && rapido->next))
            return nullptr;
        while (head != lento)
        {
            head = head->next;
            lento = lento->next;
        }
        return head;
    }
};

class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        ListNode *lento = head, *rapido = head;

        while (rapido != nullptr && rapido->next != nullptr)
        {
            lento = lento->next;
            rapido = rapido->next->next;
            if (lento == rapido)
            {
                while (head != rapido)
                {
                    head = head->next;
                    rapido = rapido->next;
                }
                return head;
            }
        }
        return nullptr;
    }
};
